server/src/sas/CompletionProvider.ts

```diff
@@ -338,6 +343,406 @@ export class CompletionProvider {
+    useSpace: boolean,
+  ): TextEdit[] {
+    // find indent text
+    const lastNotEmptyLine = this._getLastNotEmptyLine(line);
+    if (lastNotEmptyLine === undefined) {
+      return [];
+    }
+    const lastNotEmptyLineText = this.model.getLine(lastNotEmptyLine);
+    const nextLineText =
+      line < this.model.getLineCount() - 1 ? this.model.getLine(line + 1) : "";
+    const nextLineIndentText = this._getIndentText(nextLineText);
+
+    // calculate indent length
+    const curIndentLen = this._getIndentLength(lastNotEmptyLineText, tabSize);
+    const nextLineIndentInc: number | undefined =
+      this._getIndentIncrementOfNextLine(
+        lastNotEmptyLine,
+        curIndentLen,
+        tabSize,
+      );
+    if (nextLineIndentInc === undefined) {
+      return [];
+    }
+    const expectedNextLineIndentLen = curIndentLen + nextLineIndentInc;
+    const actualNextLineIndentLen = this._getIndentLength(
+      nextLineText,
+      tabSize,
+    );
+
+    if (expectedNextLineIndentLen === actualNextLineIndentLen) {
+      return [];
+    } else {
+      const expectedNextLineIndentText = this._makeIndentText(
+        expectedNextLineIndentLen,
+        useSpace,
+        tabSize,
+      );
+      return [
+        TextEdit.replace(
+          {
+            start: { line: line + 1, character: 0 },
+            end: { line: line + 1, character: nextLineIndentText.length },
+          },
+          expectedNextLineIndentText,
+        ),
+      ];
+    }
+  }
+
+  private _getIndentIncrementOfNextLine(
+    line: number,
+    curIndent: number,
+    tabSize: number,
+  ): number | undefined {
+    // find semicolon token
+    const tokens: SyntaxToken[] = this.syntaxProvider.getSyntax(line);
+    const cleanedTokens = this._cleanTokens(line, tokens);
+    if (cleanedTokens.length === 0) {
+      return 0;
+    }
+    // find patterns of "data xxx;", "proc xxx;" or "%macro xxx;"
+    const lineText = this.model.getLine(line);
+    let curIndex = cleanedTokens.length;
+    do {
+      curIndex = this._findSemicolonTokenRightToLeft(
+        line,
+        cleanedTokens,
+        curIndex - 1,
+      );
+      if (curIndex === -1) {
+        return 0;
+      }
+      const semicolonToken: SyntaxToken = cleanedTokens[curIndex];
+      const prevToken = cleanedTokens[curIndex - 1]; // curIndex must be > 0
+      const prevTokenText = this._getTokenText(
+        cleanedTokens,
+        curIndex - 1,
+        lineText,
+      )?.trim();
+      if (
+        prevTokenText === "run" ||
+        prevTokenText === "quit" ||
+        prevTokenText === "%mend"
+      ) {
+        return 0;
+      }
+      // calculate indent
+      const prevSemicolonZone: number = this.czMgr.getCurrentZone(
+        line,
+        prevToken.start,
+      );
+      if (prevSemicolonZone === CodeZoneManager.ZONE_TYPE.RESTRICTED) {
+        return undefined;
+      }
+      if (this._isDefZone(prevSemicolonZone)) {
+        return tabSize - (curIndent % tabSize);
```

Comment:
I think we should follow what VS Code natively does.
If you try a typescript editor in VS Code, it only indents to whole tab size.

Reply:
Yes, this calculation is intented to indent to whole tab size. For example, if editor.tabSize = 4, and you type only 3 spaces as the indent of a line, and then type Enter. Then INCREMENT=tabSize - (curIndent % tabSize) = 4-(3%4) = 1, The next line will be inserted 3+INCREMENT = 3+1 = 4 spaces as indent. This will promise always increase indent to whole tab size.